src/ui.cpp

```diff
@@ -639,17 +641,165 @@ void uilist::calc_data()
         } while( !assigned && hotkey != input_event() );
     }
 
-    vmax = entries.size();
+    // To figure out the layout we do two passes over the contents - first to calculate width,
+    // and only then to figure out the height, given the width restrictions.
+    // This is neccessary because we need word wrapping for useful UI,
+    // but word wrapping means that text height is affected by text width.
+
+    ImVec2 available_bounds = ImGui::GetMainViewport()->Size;
+    available_bounds.x *= 0.9; // for aesthetic reasons
+    if( desired_bounds.has_value() ) {
+        // TODO: this should never happen actually, and is a bad API
+        cataimgui::bounds b = desired_bounds.value();
+        if( b.w > 0 ) {
+            available_bounds.x = std::min( available_bounds.x, b.w );
+        }
+        if( b.h > 0 ) {
+            available_bounds.y = std::min( available_bounds.y, b.w );
+        }
+    }
 
-    ImVec2 title_size = ImVec2();
-    bool has_titlebar = !title.empty();
-    if( has_titlebar ) {
+    //
+    // TODO: calc_size does not account for items spacing
```

Comment:
Other TODOs are "nice to have" but this one is the current blocker.
I failed to figure out the exact padding of the elements, and they don't quite fit (or, alternatively, reserve too much space). Changing the spacing at runtime via ImGui debug menu should demo the issue.

Small part of the problem / minor pointer is that we create a new single-cell table that holds the entire (entire!) uilist in it for some aesthetic reasons, but the way that cell padding interacts with the other cell paddings is unobvious.

Reply:
The table is actually three columns, each containing a single cell. The middle column holds the menu. The left and right columns reserve space on the left and right respectively so that the caller can add their own UI in that space.